components/server/src/user/token-service.ts

```diff
@@ -50,13 +52,43 @@ export class TokenService implements TokenProvider {
         if (!token) {
             return undefined;
         }
+
         const aboutToExpireTime = new Date();
         aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
         if (token.expiryDate && token.expiryDate < aboutToExpireTime.toISOString()) {
+            // We attempt to get a token three times
             const { authProvider } = this.hostContextProvider.get(host)!;
+
             if (authProvider.refreshToken) {
-                await authProvider.refreshToken(user);
-                token = (await this.userDB.findTokenForIdentity(identity))!;
+                const shouldRetryRefreshTokenExchange = await getExperimentsClientForBackend().getValueAsync(
+                    "retry_refresh_token_exchange",
+                    false,
+                    {},
+                );
+                if (shouldRetryRefreshTokenExchange) {
+                    const errors: Error[] = [];
+
+                    // There is a race condition where multiple requests may each need to use the refresh_token to get a new access token.
+                    // When the `authProvider.refreshToken` is called, it will refresh the token and store it in the database.
+                    // However, the token may have already been refreshed by another request, so we need to check the database again.
+                    for (let i = 0; i < 3; i++) {
+                        try {
+                            await authProvider.refreshToken(user);
```

Comment:
@easyCZ Should we have a sleep(incl. jitter) in this loop, to ensure we don't race even more, e.g. around ~200ms?

Also in general, I think it would be easier/more stable to use a redis mutex here. But I'm happy to try this, and see if it works in gitpod.io :+1:

Reply:
Agree, redis would be better but it's also more complex. Let's see how this plays out. I'll add the backoff + jitter